### PR TITLE
as_u128 added to U256

### PR DESCRIPTION
**Pull request by alphak3y**
`as_u128` method added to `u256.sw` in order to enable downcasting from `U256` to `U128`.

### CODE REVIEW

**Comment by alphak3y** [comment]
Tests added @mohammadfawaz. Let me know if this covers it properly:
```
assert(unwrapped == ~U128::from(11, 11));
```

I guess you could also do
```
assert(unwrapped.upper == 11);
assert(unwrapped.lower == 11);
```

Of course this would fail if those field names were changed.

Thanks again!

**Review by mohammadfawaz — DISMISSED** [review]
Coverage is good. Excellent work! Thank you 🙏

**Review by otrho — APPROVED** [review]

**Comment by mohammadfawaz** [comment]
Your added tests seem to be failing. Try to run the test against a local `fuel-core` node and see what's going on.

File: sway-lib-std/src/u256.sw [hunk 1]
```diff
@@ -84,6 +84,16 @@ impl U256 {
         }
     }
 
+    /// Safely downcast to `u128` without loss of precision.
+    /// Returns Err if the number > ~128::max()
+    pub fn as_u128(self) -> Result<U128, U256Error> {
```
[the remark this answers is not in the document]
Comment (mohammadfawaz) [reply sway-lib-std/src/u256.sw:89]: Good point. It seems that this is how other methods in this library do it such as `as_u64`. I also think `TryFrom` needs generic traits, which are [nearly there!](https://github.com/FuelLabs/sway/pull/3031)

**Comment by mohammadfawaz** [comment]
@alphak3y turning this into a draft until CI passes 😄

File: test/src/e2e_vm_tests/test_programs/should_pass/stdlib/u256_test/src/main.sw [hunk 2]
```diff
@@ -95,6 +95,35 @@ fn main() -> bool {
...
+        },
+    });
+
+    let elevens = ~U256::from(0, 0, 11, 11);
+    let unwrapped = elevens.as_u128().unwrap();
+    assert(unwrapped == ~U128::from(11, 11));
+
+    let eleven = ~U256::from(0, 0, 0, 11);
+    let unwrapped = eleven.as_u128().unwrap();
+    assert(unwrapped == ~U128::from(0, 11);
```
Comment (mohammadfawaz) [inline test/src/e2e_vm_tests/test_programs/should_pass/stdlib/u256_test/src/main.sw:125]:
```suggestion
    assert(unwrapped == ~U128::from(0, 11));
```

**Comment by mohammadfawaz** [comment]
@alphak3y this was getting pretty stale so I made a few fixes. I hope you don't mind! Should be ready. I'll approve once CI passes.

**Review by mohammadfawaz — APPROVED** [review]

**Review by otrho — APPROVED** [review]

**Comment by alphak3y** [comment]
Anything is okay for me 🤝

We got around the casting by just doing it directly.

I can add more casting ability for users once we are done with the bulk of the AMM libraries.

**Pull request merged**